pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2164,6 +2168,167 @@ public boolean instanceExists(String instanceName) {
     return tableNamesWithType;
   }
 
+  /**
+   * Computes the start batch upload phase
+   *
+   * 1. Generate a batch id
+   * 2. Compute validation on the user inputs
+   * 3. Add the new lineage entry to the segment lineage metadata in the property store
+   *
+   * Update is done with retry logic along with read-modify-write block for achieving atomic update of the lineage
+   * metadata.
+   *
+   * @param tableNameWithType Table name with type
+   * @param segmentsFrom a list of segments to be merged
+   * @param segmentsTo a list of merged segments
+   * @return Bath Id
+   *
+   * @throws InvalidConfigException
+   */
+  public String startBatchUpload(String tableNameWithType, List<String> segmentsFrom, List<String> segmentsTo) {
+    // Create a batch id
+    String batchId = UUID.randomUUID().toString();
+
+    // Check that segmentsTo is not empty.
+    Preconditions.checkArgument(!segmentsTo.isEmpty(), String
+        .format("'segmentsTo' cannot be empty (tableName = '%s', segmentsFrom = '%s', segmentsTo = '%s'",
+            tableNameWithType, segmentsFrom, segmentsTo));
+
+    // Check that all the segments from 'segmentsFrom' exist in the table
+    Set<String> segmentsForTable = new HashSet<>(getSegmentsFor(tableNameWithType));
+    Preconditions.checkArgument(segmentsForTable.containsAll(segmentsFrom), String.format(
+        "Not all segments from 'segmentsFrom' are available in the table. (tableName = '%s', segmentsFrom = '%s', "
+            + "segmentsTo = '%s', segmentsFromTable = '%s')", tableNameWithType, segmentsFrom, segmentsTo,
+        segmentsForTable));
+
+    // Check that all the segments from 'segmentTo' does not exist in the table.
+    Preconditions.checkArgument(Collections.disjoint(segmentsForTable, segmentsTo), String.format(
+        "Any segments from 'segmentsTo' should not be available in the table at this point. (tableName = '%s', "
+            + "segmentsFrom = '%s', segmentsTo = '%s', segmentsFromTable = '%s')", tableNameWithType, segmentsFrom,
+        segmentsTo, segmentsForTable));
+
+    try {
+      final List<String> finalSegmentsFrom = segmentsFrom;
+      DEFAULT_RETRY_POLICY.attempt(() -> {
+        // Fetch the segment lineage metadata
+        ZNRecord segmentLineageZNRecord =
+            SegmentLineageAccessHelper.getSegmentLineageZNRecord(_propertyStore, tableNameWithType);
+        SegmentLineage segmentLineage;
+        int expectedVersion = -1;
+        if (segmentLineageZNRecord == null) {
+          segmentLineage = new SegmentLineage(tableNameWithType);
+        } else {
+          segmentLineage = SegmentLineage.fromZNRecord(segmentLineageZNRecord);
+          expectedVersion = segmentLineageZNRecord.getVersion();
+        }
+
+        // Check that the batchId doesn't exists in the segment lineage
+        Preconditions.checkArgument(segmentLineage.getLineageEntry(batchId) == null,
+            String.format("BatchId (%s) already exists in the segment lineage.", batchId));
```

Comment:
Don't directly throw exception as it will abort the retry.
You can keep generating `batchId` until get a new one (in most case there is no collision)
```suggestion
        String batchId = SegmentLineageUtils. generateLineageEntryId();
        while (segmentLineage.getLineageEntry(batchId) != null) {
          batchId = SegmentLineageUtils. generateLineageEntryId();
        }
```

Reply:
The reason why I generate `batchId` outside of retry block is that it's a bit hard to pass some information from retry block to the caller based on the current `RetryPolicy` interface. For now, let's not deal with the id collision. Throwing exception should be fine. Once this part shows up frequently (likely not because it's extremely hard to observe UUID collision), we can deal with this by improving retry policy interface to allow pass information back to the caller.